### DISCUSSION TITLE
Vector memory exhausted when running methylKit

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1137] [Q&A] [answered]
I am running `methylKit` but ran into an error:

![Screen Shot 2021-03-07 at 3 40 48 PM](https://user-images.githubusercontent.com/22335838/110259243-7ebd0380-7f5b-11eb-9db3-bc52af520162.png)

I thought I could use the R Studio server to run the code with more memory. When I tried opening each server, it seemed like they were frozen! I was unable to interact with R Studio through the server on roadrunner (I tried clicking things and nothing happened), and the R Studio server on emu was stuck for ~10 minutes trying to terminate R. 

Any suggestions for using the servers, or any non-server workarounds/suggestions for processing my files?

**Comment by yaaminiv** [comment]
For reference, here's my script: https://github.com/RobertsLab/project-oyster-oa/blob/master/code/Haws/04-methylKit.Rmd#L105

In a previous iteration of the script, I was able to process my files without errors: https://yaaminiv.github.io/Hawaii-Gigas-Methylation-Analysis-Part7/

I'm testing the `normalizeCoverage` function within `methylKit`, which I think is causing the vector memory issue.

**Reply by sr320** [reply]
note I will try to reproduce locally

**Reply by sr320** [reply]
some odd filenames `zr3644_1_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov`

**Reply by sr320** [reply]
@yaaminiv can you give me access to bams? I think this filetype does influence memory; deduplicated and sorted

**Reply by sr320** [reply]
found the bams

**Reply by yaaminiv** [reply]
Sorry for the delay! Odd filenames are because I trimmed 3 times (2 trimgalore adapter removal, 1 poly-G tail removal) and I never bothered to rename

**Reply by yaaminiv** [reply]
>  can you give me access to bams? I think this filetype does influence memory; deduplicated and sorted

Speaking of bams, I thought when we were working with the coral data that we decided to process the merged cov files with methylkit, and not bams, because they were already merged. Wouldn't the bams also be larger?

**Comment by kubu4** [comment]
> When I tried opening each server, it seemed like they were frozen!

Can you please clarify this point? What do you mean "each server"? Meaning Emu and Roadrunner?

I logged into RStudio Server on both computers without issue (with my user account). This might suggest that you already had an existing RStudio Server session open from previous usage. Possibly tied into the default behavior where RStudio attempts to load the last used RStudio session. If the previous session had a lot of variables/dataframes (and, those things contained large amounts of data), it might cause the session to freeze and/or take a long time to open as it loads all of those things into memory.

I'll see if I can figure out how to disable this "feature" for you. Also, which user account are you using?

**Reply by kubu4** [reply]
Alrighty, you can now login to RStudio Server on roadrunner with the srlab account without it freezing. Plus, it now has a clean session.

I *think* I resolved this by moving the `~/.rstudio` directory to `~/.rstudio-backup`. This, effectively, removes the `~/.rstudio` directory, which is where session info is saved/loaded. 

Got that from this [rstudio GitHub Issue](https://github.com/rstudio/rstudio/issues/1633). However, it seems like this issue of automatically loading the previous RStudio session has been fixed in a more recent version of Rstudio Server. I'll explore upgrading so we don't have to mess around with this.

**Reply by kubu4** [reply]
Finally, I'd like to add, I don't know if this will fix your memory issue. But, it does allow you to use RStudio Server. Try running your script again and let us know how it goes...

**Reply by yaaminiv** [reply]
> Can you please clarify this point? What do you mean "each server"? Meaning Emu and Roadrunner?

Yes I meant Emu and Roadrunner!

> I logged into RStudio Server on both computers without issue (with my user account). 

Do I have a separate server account? I was using the srlab one, which is trying to resume the last opened R session

**Reply by yaaminiv** [reply]
Never mind, just saw that you restarted the srlab account on Roadrunner. Cloning my repo now and will try and run shortly

**Reply by yaaminiv** [reply]
> Alrighty, you can now login to RStudio Server on roadrunner with the srlab account without it freezing. Plus, it now has a clean session.

I think I may have screwed it up...I ran a bash command within my R Markdown script and now the server keeps freezing on the srlab account on Roadrunner. Is there a way I can reset it?

**Reply by kubu4** [reply]
SSH into roadrunner and do this:

> resolved this by moving the ~/.rstudio directory to ~/.rstudio-backup.

**Reply by yaaminiv** [reply]
> moving the ~/.rstudio directory to ~/.rstudio-backup.

I wasn't able to move the directory itself to  ~/.rstudio-backup

![Screen Shot 2021-03-08 at 1 04 58 PM](https://user-images.githubusercontent.com/22335838/110381838-e54f2980-800e-11eb-9f0b-d325b459a767.png)

I moved the individual monitored and sessions directories:

![Screen Shot 2021-03-08 at 1 04 19 PM](https://user-images.githubusercontent.com/22335838/110381780-cc467880-800e-11eb-9a88-3b522e76e3a3.png)

When I open the R Studio server, it opens my previous script and is still frozen. Is there a different way I should be moving the directories?

**Reply by kubu4** [reply]
You have to delete the previous backup first (or, move to a differently named destination).

I went ahead and did this.

Try RStudio Server again.

**Reply by yaaminiv** [reply]
Still frozen on my script when I log into the srlab account

**Reply by kubu4** [reply]
You can load it now. I restarted RStudio Server. But, I'd refer to @sr320 's reply below RE: Mox.

**Reply by yaaminiv** [reply]
With respect to the original vector memory issue, I was able to run `filterByCoverage` + `normalizeCoverage` on the R Studio server (finally) without running into a memory issue. So at least that's good

**Comment by sr320** [comment]
is mox a viable option here? I have always had issues with R Server.

**Reply by yaaminiv** [reply]
I've never tried running an R script on mox, but if there's a way I could try it?

**Reply by kubu4** [reply]
At this point, that'd be my suggestion as well - use R from command line instead. Might not even need Mox.

**Reply by sr320** [reply]
I presume the original error in question here is something you got from running on your local machine?

**Reply by kubu4** [reply]
It's basically the same thing as running an R script from any command line:

`R name_of_R-script.R`

So, inside an SBATCH script for Mox, you'd have all of your header junk specifying run time, directory, etc. Then, you'd have the line above which would call and execute an R script that you have somewhere.

**Reply by yaaminiv** [reply]
I was running it on genefish because I figured it would take a while and wanted to be able to do other things on my local machine, but I can try my laptop.

R studio server is unfrozen, but I'm running into different errors since I think I'm running a more updated version of methylKit. I'll see if I can troubleshoot it and get to the command that caused my initial error

**Reply by kubu4** [reply]
Alternatively, if you edit the top line of the SBATCH file to point to the R installation, then you can put the contents of your R script in the SBATCH file, just like you'd normally do:

Change this line:

`#!/bin/bash`

to

`#!/gscratch/srlab/programs/R-3.6.2/bin/R`

**Comment by sr320** [comment]
My 2 cents is make sure it goes error free on locally on your laptop before
moving off to new platform (noting I get a earlier methylkit error trying
to reproduce your code.

**Reply by yaaminiv** [reply]
definitely. I'm trying to see if I can get error free on the R Studio Server, but I think I froze it again. I'll try on my local machine with a subset of files to check, then move to mox for the full dataset

**Comment by yaaminiv** [comment, marked as answer]
Update: was able to run `filterByCoverage` and `normalizeCoverage` with the Manchester data (8 samples) without running into a vector memory issue. So I think R was having a problem with the amount of data being processed at once (24 Hawaii samples).

Still working on debugging script with Manchester data, before running the same code on the Hawaii data on Mox.
